### PR TITLE
new(cmd/driver): support user provided headers option for driver download

### LINKED ISSUE #426: Option to add headers to the http request of falcoctl driver install command
**What would you like to be added**:

According to my understanding, it is currently not possible to add headers in the http GET request which downloads a pre-built driver, using the falcoctl driver install command. This functionality should be added again.

**Why is this needed**:
This had been possible until the migration from script to falcoctl based download via the additional curl options via an environment variable.

### PR DESCRIPTION
**Pull request by toamto94**
**What type of PR is this?**

> Uncomment one (or more) `/kind <>` lines:

> /kind bug

> /kind cleanup

> /kind design

> /kind documentation

> /kind failing-test

/kind feature

> /kind flaky-test

**Any specific area of the project related to this PR?**

> Uncomment one (or more) `/area <>` lines:

> /area library

/area cli

> /area tests

> /area examples

**What this PR does / why we need it**:
Added option to add headers in the http GET request for the driver download

**Which issue(s) this PR fixes**:
Option had been possible via environment variables in the script based download but is not possible anymore since the falcoctl migration

Fixes #426

**Special notes for your reviewer**:
example flag (for testing):
--http-headers="x-emc-namespace: default,Proxy-Authenticate: Basic"

### CODE REVIEW

**Comment by FedeDP** [comment]
Thanks for this PR! I'll give it a look tomorrow!
/assign

File: cmd/driver/install/install.go [hunk 1]
```diff
@@ -54,8 +55,9 @@ func NewDriverInstallCmd(ctx context.Context, opt *options.Common, driver *optio
 		Common: opt,
 		Driver: driver,
 		// Defaults to downloading or building if needed
-		Download: true,
-		Compile:  true,
+		Download:    true,
+		Compile:     true,
+		HTTPHeaders: "",
```
Comment (FedeDP) [inline cmd/driver/install/install.go:60]: We have a `driverDownloadOptions` struct to hold http related things; can you put this in there?

File: cmd/driver/install/install.go [hunk 2]
```diff
@@ -92,6 +94,10 @@ func NewDriverInstallCmd(ctx context.Context, opt *options.Common, driver *optio
 			"(e.g. '#1 SMP PREEMPT_DYNAMIC Debian 6.1.38-2 (2023-07-27)')")
 	cmd.Flags().BoolVar(&o.InsecureDownload, "http-insecure", false, "Whether you want to allow insecure downloads or not")
 	cmd.Flags().DurationVar(&o.HTTPTimeout, "http-timeout", 60*time.Second, "Timeout for each http try")
+	cmd.Flags().StringVar(&o.HTTPHeaders, "http-headers",
+		"",
+		"Optional comma-separated list of headers for the http GET request."+
```
Comment (FedeDP) [inline cmd/driver/install/install.go:99]: Can you also add a line with an example here?

**Comment by FedeDP** [comment]
To fix linting: make sure to split the line:
```
driverdistro.Download(ctx, d, o.Printer.WithWriter(&buf), kr, o.Driver.Name, o.Driver.Type, o.Driver.Version, o.Driver.Repos, o.HTTPHeaders)
```
Into multiple lines ;)

File: pkg/driver/distro/distro.go [hunk 3]
```diff
@@ -218,6 +219,17 @@ func Download(ctx context.Context,
 		printer.Logger.Info("Trying to download a driver.", printer.Logger.Args("url", url))
 
 		req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
+		if httpHeaders != "" {
```
Comment (FedeDP) [inline pkg/driver/distro/distro.go:222]: Also, i'd move this check below the `if err != nil` check: no need to set headers to a possibly nil `req`.

**Review by FedeDP — APPROVED** [review]
/approve

**Comment by FedeDP** [comment]
Can you run `make fmt` to fix linting? Thanks!

File: cmd/driver/install/install.go [hunk 4]
```diff
@@ -36,13 +36,14 @@ import (
...
 	HTTPTimeout      time.Duration
+	HTTPHeaders      string
 }
 
 type driverInstallOptions struct {
 	*options.Common
 	*options.Driver
-	Download            bool
-	Compile             bool
+	Download: true,
```
Comment (FedeDP) [inline cmd/driver/install/install.go:45]: `bool` ?

**Comment by toamto94** [comment]
Ah I am sorry! I have a meeting now and will fix it afterwords. Gesendet mit der mobilen Mail AppAm 09.02.24 um 10:42 schrieb Federico Di Pierro
                
            
            
                
                    Von: "Federico Di Pierro" ***@***.***>Datum: 9. Februar 2024An: "falcosecurity/falcoctl" ***@***.***>Cc: "Mention" ***@***.***>,"Tom Müller" ***@***.***>Betreff: Re: [falcosecurity/falcoctl] Fb header options (PR #428)
                
                
@FedeDP commented on this pull request.



In cmd/driver/install/install.go:
>  }
 
 type driverInstallOptions struct {
 	*options.Common
 	*options.Driver
-	Download            bool
-	Compile             bool
+	Download: true,

bool ?

**Comment by FedeDP** [comment]
I'd say we are quite ready; you just miss to run `make fmt` and perhaps i'd squash everything in a single commit :)

**Review by toamto94 — COMMENTED** [review]
My IDE is messing around, I am sorry for the confusion

**Comment by toamto94** [comment]
> I'd say we are quite ready; you just miss to run `make fmt` and perhaps i'd squash everything in a single commit :)

Alright, let me get things right now. Sorry for the confusion :D

**Comment by FedeDP** [comment]
No problem :) It can be quite cumbersome first times!

**Comment by FedeDP** [comment]
Oh you need to also bump helper message here: https://github.com/falcosecurity/falcoctl/blob/main/cmd/driver/install/install_test.go#L29
Sorry, forgot about it!

**Comment by toamto94** [comment]
> Oh you need to also bump helper message here: https://github.com/falcosecurity/falcoctl/blob/main/cmd/driver/install/install_test.go#L29 Sorry, forgot about it!

Ok no problem!

**Comment by toamto94** [comment]
ah need to sign of. sec

**Comment by toamto94** [comment]
Aah wait let me fix this

**Comment by toamto94** [comment]
Fighting with git is always fun :D

**Comment by toamto94** [comment]
Failed. What exactly needs to be done there @FedeDP

**Comment by toamto94** [comment]
Don't seem to work hm. Any idea what I am doing wrong?

**Comment by FedeDP** [comment]
You can quickly locally test using `go test ./cmd/driver/...` ; am not sure about what's failing though :(

**Comment by toamto94** [comment]
> You can quickly locally test using `go test ./cmd/driver/...` ; am not sure about what's failing though :(

Ah alright! Thx for the tip

**Comment by FedeDP** [comment]
```
  --http-headers string     Optional comma-separated list of headers for the http GET request (e.g. --http-headers="x-emc-namespace: default,Proxy-Authenticate: Basic"). Not necessary if default repo is used
      --http-insecure           Whether you want to allow insecure downloads or not
```
`--http-headers` should be above `--http-insecure` :D

File: cmd/driver/install/install_test.go [hunk 5]
```diff
@@ -37,6 +37,7 @@ Flags:
       --download                Whether to enable download of prebuilt drivers (default true)
   -h, --help                    help for install
       --http-insecure           Whether you want to allow insecure downloads or not
+      --http-headers string     Optional comma-separated list of headers for the http GET request (e.g. --http-headers='x-emc-namespace: default,Proxy-Authenticate: Basic'). Not necessary if default repo is used
```
Comment (FedeDP) [inline cmd/driver/install/install_test.go:40]: Here, `--http-headers` should be before `--http-insecure`.

Comment (toamto94) [reply cmd/driver/install/install_test.go:39]: Awesome!! Thank you very much! :)

**Review by FedeDP — APPROVED** [review]
/approve

**Pull request merged**